apiserver/tools.go

```diff
@@ -20,20 +20,121 @@ import (
+	coreerrors "github.com/juju/juju/core/errors"
 	coremodel "github.com/juju/juju/core/model"
 	"github.com/juju/juju/core/objectstore"
 	"github.com/juju/juju/core/semversion"
+	agentbinaryerrors "github.com/juju/juju/domain/agentbinary/errors"
 	"github.com/juju/juju/environs"
 	"github.com/juju/juju/environs/simplestreams"
 	envtools "github.com/juju/juju/environs/tools"
+	internalerrors "github.com/juju/juju/internal/errors"
 	jujuhttp "github.com/juju/juju/internal/http"
+	"github.com/juju/juju/internal/services"
 	"github.com/juju/juju/internal/tools"
 	"github.com/juju/juju/rpc/params"
 	"github.com/juju/juju/state"
 	"github.com/juju/juju/state/binarystorage"
 	"github.com/juju/juju/state/stateenvirons"
 )
 
+// AgentBinaryStore is an interface that provides the ability to store new agent
+// binaries into a store for the controller or model.
+type AgentBinaryStore interface {
+	// AddAgentBinaryWithSHA256 adds a new agent binary to the object store and
+	// saves its metadata to the database.
+	// The following errors can be returned:
+	// - [coreerrors.NotSupported] if the architecture is not supported.
+	// - [agentbinaryerrors.AlreadyExists] if an agent binary already exists for
+	// this version and architecture.
+	// - [agentbinaryerrors.ObjectNotFound] if there was a problem referencing
+	// the agent binary metadata with the previously saved binary object.
+	// This error should be considered an internal problem. It is discussed here
+	// to make the caller aware of future problems.
+	// - [coreerrors.NotValid] if the agent version is not valid or the SHA256
+	// hash doesn't match the generated hash.
+	AddAgentBinaryWithSHA256(
+		context.Context,
+		io.Reader,
+		coreagentbinary.Version,
+		int64,
+		string,
+	) error
+}
+
+// AgentBinaryStoreGetter is a deferred type that can be used to get an
+// AgentBinaryStore at exactly the time it is needed. This allows for the
+// context aware answers to be made.
+type AgentBinaryStoreGetter func(*http.Request) (AgentBinaryStore, error)
+
+type BlockChecker interface {
+	// ChangeAllowed checks if change block is in place.
+	// Change block prevents all operations that may change
+	// current model in any way from running successfully.
+	ChangeAllowed(context.Context) error
+}
+
+// DomainServicesGetter describes a type that can be used for getting
+// [services.DomainServices] from a given context that comes from a http
+// request.
+type DomainServicesGetter func(ctx context.Context) (services.DomainServices, error)
+
+// BlockCheckerGetterForServices returns a [BlockCheckerGetter] that is
+// constructed from the supplied context.
+func BlockCheckerGetterForServices(servicesGetter DomainServicesGetter) func(context.Context) (BlockChecker, error) {
+	return func(ctx context.Context) (BlockChecker, error) {
+		svc, err := servicesGetter(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		return common.NewBlockChecker(svc.BlockCommand()), nil
+	}
+}
```

Comment:
This is duplicated in `apiserver/common/block.go` and the latter is unused. Given a block checker is going to be used elsewhere, the one in common should be used and this one deleted.

Reply:
Agreed, but as per comments above from a testing dependency point of view we need to define it here otherwise we have to mock the implementation of BlockChecker it self in an unrelated component.